File: python/problem1392.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        N = len(s)
        next = [-1 for _ in range(N + 1)]
        k = -1
        j = 0
        while j < N:
            if k == -1 or s[j] == s[k]:
                k += 1
                j += 1
                next[j] = k
            else:
                k = next[k]
        return s[0 : next[N]]

File: Java/problem1392.java
class Solution {
    
    public String longestPrefix(String s) {
        char[] chars = s.toCharArray();
        int L = chars.length;
        int[] next = new int[L + 1];
        next[0] = -1;
        int k = -1, j = 0;
        while (j < chars.length) {
            if (k == -1 || chars[j] == chars[k]) {
                next[++j] = ++k;
            }
            else {
                k = next[k];
            }
        }
        return next[L] == -1 ? "" : s.substring(0, next[L]);
    }
}
